### ISSUE TITLE
pods can no longer reach network/internet

### ISSUE THREAD

**Issue by mhegreberg** [rancher/rke2#5929]
**Environmental Info:**
RKE2 Version: 

rke2 version v1.26.15+rke2r1 (a413a7fe6321b73971e66b6b8eb87900a40de559)
go version go1.21.8 X:boringcrypto

Node(s) CPU architecture, OS, and Version: 

Linux rke2-prod1-worker1 6.1.0-20-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.85-1 (2024-04-11) x86_64 GNU/Linux

Cluster Configuration:

3 Server nodes, 2 worker nodes

**Describe the bug:**

pods cannot reach out to internet.

**Steps To Reproduce:**

- Installed RKE2 via rancher a few months ago, everything was working fine.
- have updated k8s a few times via rancher
- recently noticed that pods cannot reach out to network/internet, and are having other networking issues.
- Ive even disabled node firewalls to verify that's not the issue

**Expected behavior:**

pods can reach out to network/internet for resources

**Actual behavior:**

pods fail with various types of "resource unavailable" type erros

**Additional context / logs:**

I get these error every time I run a kubectl command:
![image](https://github.com/rancher/rke2/assets/11844475/3d7b7264-dab7-40c4-8d62-785429e439d1)

**Comment by brandond** [comment]
Have you checked the logs on your server nodes? Are there any pods in error state? The messages suggest that something is going on with your metrics-server deployment, at the very least.

**Comment by mhegreberg** [comment]
I don't see anything super interesting when I run `journalctl -u rke2-server` I saw a few entries that were kind of unusual, but I don't really suspect them. sholuld I be looking for something specific?

```
May 20 08:36:08 rke2-prod1-server1 rke2[855]: time="2024-05-20T08:36:08-07:00" level=info msg="Handling backend connection request [rke2-prod1-worker1]"
May 20 08:39:03 rke2-prod1-server1 rke2[855]: 2024/05/20 08:39:03 ERROR: [transport] Client received GoAway with error code ENHANCE_YOUR_CALM and debug data equal to ASCII "too_many_pings".
May 20 08:54:43 rke2-prod1-server1 rke2[855]: 2024/05/20 08:54:43 ERROR: [transport] Client received GoAway with error code ENHANCE_YOUR_CALM and debug data equal to ASCII "too_many_pings".
May 20 08:58:53 rke2-prod1-server1 rke2[855]: 2024/05/20 08:58:53 ERROR: [transport] Client received GoAway with error code ENHANCE_YOUR_CALM and debug data equal to ASCII "too_many_pings".
May 20 09:31:14 rke2-prod1-server1 rke2[855]: 2024/05/20 09:31:14 ERROR: [transport] Client received GoAway with error code ENHANCE_YOUR_CALM and debug data equal to ASCII "too_many_pings".
May 20 09:51:14 rke2-prod1-server1 rke2[855]: 2024/05/20 09:51:14 ERROR: [transport] Client received GoAway with error code ENHANCE_YOUR_CALM and debug data equal to ASCII "too_many_pings".
```

rke2-metrics-server seems to be running. here are the logs from it:
```
% kubectl -n kube-system logs rke2-metrics-server-d976688df-rhzwk                                                                                                                                        ~
E0520 13:12:05.877518  176632 memcache.go:287] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E0520 13:12:05.884827  176632 memcache.go:287] couldn't get resource list for custom.metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E0520 13:12:05.900417  176632 memcache.go:121] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E0520 13:12:05.911091  176632 memcache.go:121] couldn't get resource list for custom.metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E0520 13:12:05.924940  176632 memcache.go:121] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E0520 13:12:05.936718  176632 memcache.go:121] couldn't get resource list for custom.metrics.k8s.io/v1beta1: the server is currently unable to handle the request
I0520 15:36:56.438009       1 secure_serving.go:267] Serving securely on [::]:10250
I0520 15:36:56.438170       1 requestheader_controller.go:169] Starting RequestHeaderAuthRequestController
I0520 15:36:56.441364       1 shared_informer.go:240] Waiting for caches to sync for RequestHeaderAuthRequestController
W0520 15:36:56.441426       1 shared_informer.go:372] The sharedIndexInformer has started, run more than once is not allowed
I0520 15:36:56.438238       1 configmap_cafile_content.go:201] "Starting controller" name="client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file"
I0520 15:36:56.441656       1 shared_informer.go:240] Waiting for caches to sync for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I0520 15:36:56.438252       1 dynamic_serving_content.go:131] "Starting controller" name="serving-cert::/tmp/apiserver.crt::/tmp/apiserver.key"
I0520 15:36:56.438263       1 tlsconfig.go:240] "Starting DynamicServingCertificateController"
I0520 15:36:56.438371       1 configmap_cafile_content.go:201] "Starting controller" name="client-ca::kube-system::extension-apiserver-authentication::client-ca-file"
I0520 15:36:56.441713       1 shared_informer.go:240] Waiting for caches to sync for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I0520 15:36:56.542066       1 shared_informer.go:247] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I0520 15:36:56.542141       1 shared_informer.go:247] Caches are synced for RequestHeaderAuthRequestController
I0520 15:36:56.543281       1 shared_informer.go:247] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::client-ca-file

```

**Comment by brandond** [comment]
The metrics errors are odd, but probably unrelated to whatever's going on with your outbound traffic. Are all of the CNI pods functioning properly?

**Comment by mhegreberg** [comment]
they appear to be running:
```
NAMESPACE                  NAME                                                          READY   STATUS      RESTARTS        AGE
calico-system              calico-kube-controllers-67765866fb-227wg                      1/1     Running     1 (5h8m ago)    33d
calico-system              calico-node-5fjlv                                             1/1     Running     1 (5h9m ago)    33d
calico-system              calico-node-hl72n                                             1/1     Running     1 (5h10m ago)   33d
calico-system              calico-node-kfxd2                                             1/1     Running     1 (5h10m ago)   33d
calico-system              calico-node-mt5fw                                             1/1     Running     1 (5h8m ago)    33d
calico-system              calico-node-thj2q                                             1/1     Running     2 (5h8m ago)    33d
calico-system              calico-typha-677656586b-bh868                                 1/1     Running     1 (5h10m ago)   33d
calico-system              calico-typha-677656586b-jwcbb                                 1/1     Running     1 (5h8m ago)    20d
calico-system              calico-typha-677656586b-qfjqg                                 1/1     Running     1 (5h8m ago)    33d
```

I don't see any errors in the logs, but I did see this in the typha logs?

```
2024-05-20 16:17:15.808 [INFO][7] watchercache.go 181: Full resync is required ListRoot="/calico/ipam/v2/host/"
2024-05-20 17:09:42.714 [INFO][7] watchercache.go 125: Watch error received from Upstream ListRoot="/calico/ipam/v2/host/" error=too old resource version: 3411 (54910102)
2024-05-20 17:09:43.213 [INFO][7] watchercache.go 181: Full resync is required ListRoot="/calico/ipam/v2/host/"
2024-05-20 17:41:04.361 [INFO][7] watchercache.go 125: Watch error received from Upstream ListRoot="/calico/ipam/v2/host/" error=too old resource version: 4274 (54910102)
2024-05-20 17:41:04.851 [INFO][7] watchercache.go 181: Full resync is required ListRoot="/calico/ipam/v2/host/"
2024-05-20 18:25:43.746 [INFO][7] watchercache.go 125: Watch error received from Upstream ListRoot="/calico/ipam/v2/host/" error=too old resource version: 4274 (54910102)
2024-05-20 18:25:44.242 [INFO][7] watchercache.go 181: Full resync is required ListRoot="/calico/ipam/v2/host/"
2024-05-20 19:02:04.020 [INFO][7] watchercache.go 125: Watch error received from Upstream ListRoot="/calico/ipam/v2/host/" error=too old resource version: 1246 (54910102)
2024-05-20 19:02:04.508 [INFO][7] watchercache.go 181: Full resync is required ListRoot="/calico/ipam/v2/host/"
2024-05-20 19:36:19.218 [INFO][7] watchercache.go 125: Watch error received from Upstream ListRoot="/calico/ipam/v2/host/" error=too old resource version: 4274 (54910102)
2024-05-20 19:36:19.718 [INFO][7] watchercache.go 181: Full resync is required ListRoot="/calico/ipam/v2/host/"
2024-05-20 20:32:58.042 [INFO][7] watchercache.go 125: Watch error received from Upstream ListRoot="/calico/ipam/v2/host/" error=too old resource version: 3411 (54910102)
```


it's possible this started after updating to v1.26.15+rke2r1. I did the update and went on leave shortly after. this cluster is still being evaluated, so not a lot of important workloads are on it yet.

**Comment by manuelbuil** [comment]
`pods cannot reach out to internet.` what's the error you see?

**Comment by mhegreberg** [comment]
@manuelbuil I'm seeing "resource temporarily unavailable" errors from pods running applications anytime they try to reach out to a DB server, the internet, logging server, etc.

here is an example from a dotnet app:
```
---> System.Net.Http.HttpRequestException: Resource temporarily unavailable (login.microsoftonline.com:443)
 ---> System.Net.Sockets.SocketException (11): Resource temporarily unavailable
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at System.Net.Sockets.Socket.<ConnectAsync>g__WaitForConnectWithCancellation|285_0(AwaitableSocketAsyncEventArgs saea, ValueTask connectTask, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.ConnectToTcpHostAsync(String host, Int32 port, HttpRequestMessage initialRequest, Boolean async, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
```

I also get similar errors every time I run a kubectl command:
```
E0521 09:56:01.466006     798 memcache.go:287] couldn't get resource list for custom.metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E0521 09:56:01.479003     798 memcache.go:287] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E0521 09:56:01.525483     798 memcache.go:121] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
```

I've also noticed that trying to view any pods/deployments from the rancher interface takes many minutes, instead of the moment they took before, and take with other clusters we have.

I can ping these resources from the host nodes, so I know that its not an external network issue.

**Comment by brandond** [comment]
Can you confirm what address these things are resolving to from within pods? It sounds more like you're not even hitting the intended destination, but rather some other endpoint that's returning an error.

**Comment by mhegreberg** [comment]
I think you're right, it's something to do with coredns. I spun up a busybox pod and can ping 8.8.8.8 but not google.com.

I see the coredns IP is in the resolv.conf of a pod:
```
%kubectl --namespace default exec -ti dnsutils -- cat /etc/resolv.conf                                                                                                                             ~
E0521 12:10:29.083269    2861 memcache.go:287] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E0521 12:10:29.096685    2861 memcache.go:287] couldn't get resource list for custom.metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E0521 12:10:29.171591    2861 memcache.go:121] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E0521 12:10:29.181595    2861 memcache.go:121] couldn't get resource list for custom.metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E0521 12:10:29.194355    2861 memcache.go:121] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E0521 12:10:29.205501    2861 memcache.go:121] couldn't get resource list for custom.metrics.k8s.io/v1beta1: the server is currently unable to handle the request
search default.svc.cluster.local svc.cluster.local cluster.local
nameserver 10.43.0.10
options ndots:5
```
but if I try to run any  lookups I get get "no server can be reached"


```
%kubectl --namespace default exec -i -t dnsutils -- nslookup kubernetes.default                                                                                                                    ~
E0521 12:11:13.023656    2897 memcache.go:287] couldn't get resource list for custom.metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E0521 12:11:13.024445    2897 memcache.go:287] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E0521 12:11:13.034527    2897 memcache.go:121] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E0521 12:11:13.044874    2897 memcache.go:121] couldn't get resource list for custom.metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E0521 12:11:13.055507    2897 memcache.go:121] couldn't get resource list for metrics.k8s.io/v1beta1: the server is currently unable to handle the request
E0521 12:11:13.065586    2897 memcache.go:121] couldn't get resource list for custom.metrics.k8s.io/v1beta1: the server is currently unable to handle the request
;; connection timed out; no servers could be reached

command terminated with exit code 1
(1)%                                         
```

as far as I can tell coredns is running, and the service is up, but requests aren't reaching it.

**Comment by brandond** [comment]
Can you make lookups against the coredns pod IP, or does that fail? That should indicate if all CNI traffic is affected, or just traffic to the service clusterip.

**Comment by mhegreberg** [comment]
From inside a pod:
```
# nslookup google.com
;; connection timed out; no servers could be reached

# nslookup google.com 8.8.8.8
Server:         8.8.8.8
Address:        8.8.8.8#53

Non-authoritative answer:
Name:   google.com
Address: 142.250.68.110


# nslookup google.com 10.42.247.75
;; connection timed out; no servers could be reached

#
```

I also noticed that I cannot ping the coredns pods, either through their direct IP or through the service IP. 

it seems we're getting closer!

I went ahead and restarted the rke2-coredns-rke2-coredns deployment to see if that would help, but same issue

**Comment by brandond** [comment]
It sounds like something is blocking or dropping CNI traffic between nodes, however it's not clear if this is affecting both TCP and UDP or just UDP. Have you made any changes to policies on external firewalls? Can you confirm that any local iptables-based firewalls (firewalld/ufw) are disabled on these nodes?

**Comment by manuelbuil** [comment]
Another posibility is that you could be hitting the vxlan kernel issue: https://github.com/rancher/rke2/issues/1541#issuecomment-895119362

**Comment by mhegreberg** [comment]
I turned off ufw on all the nodes at the beginning of troubleshooting.

looking through iptables I was able to discover the issue. for some reason all the inter-node traffic was being routed through a different network interface (the one for the SAN network) as opposed to the one that connects to the other pods and network at large.

not sure what caused the change. I was able to temporarily disable the second NIC.

At this point I think I'll do more research as to why the host node routing tables changed. If I find that it was a direct issue with rke2, and not with out configuration, I'll reopen this.

thanks for all your help!
